runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java

```diff
@@ -214,21 +213,22 @@ State waitUntilFinish(
     MonitoringUtil monitor = new MonitoringUtil(projectId, dataflowOptions.getDataflowClient());
 
     long lastTimestamp = 0;
-    BackOff backoff =
-        duration.getMillis() > 0
-            ? new AttemptAndTimeBoundedExponentialBackOff(
-                MESSAGES_POLLING_ATTEMPTS,
-                MESSAGES_POLLING_INTERVAL,
-                duration.getMillis(),
-                AttemptAndTimeBoundedExponentialBackOff.ResetPolicy.ATTEMPTS,
-                nanoClock)
-            : new AttemptBoundedExponentialBackOff(
-                MESSAGES_POLLING_ATTEMPTS, MESSAGES_POLLING_INTERVAL);
+    FlexibleBackoff backoff = FlexibleBackoff.of()
+        .withMaxRetries(MESSAGES_POLLING_RETRIES)
+        .withInitialBackoff(MESSAGES_POLLING_INTERVAL);
+    if (duration.isLongerThan(Duration.ZERO)) {
```

Comment:
`withMaxCumulativeBackoff` could have the logic that duration <= 0 means no limit. Then the if statement can move into the utility.